### PR TITLE
Component table

### PR DESCRIPTION
**Pull request by xzdry**

### CODE REVIEW

File: src/components/Table.tsx [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+import { useMemo, useRef, useState } from "react";
+import { sortBy } from "lodash-es";
+import {
+  LIST_ITEM_EXP,
+  LIST_ITEM_INDEX_EXP,
+  ModuleRenderer,
+} from "@sunmao-ui/runtime";
+
+const TableStateSchema = Type.Object({
+  selectedRows: Type.Optional(Type.Array(Type.Any())),
```
Comment (tanbowensg) [inline src/components/Table.tsx]: It should not be optional. It should be `[]` when no rows are selected.

File: src/components/Table.tsx [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    displayName: "Table",
+  },
+  spec: {
+    properties: TablePropsSchema,
+    state: TableStateSchema,
+    methods: {
+      print: Type.String(),
+      deleteSelected: Type.String(),
+      deleteRows: Type.String(),
+    },
```
Comment (tanbowensg) [inline src/components/Table.tsx]: should delete these methods

File: src/generated/types/Table.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}))
+
+export const ColumnSchema = Type.Object({
+  title: Type.String(),
+  dataIndex: Type.String(),
+  sorter: Type.Optional(Type.Boolean()),
+  filter:Type.Optional(Type.Boolean()),
+  sortDirections: Type.Optional(Type.Array(StringUnion(["descend" ,"ascend"]))),
+  defaultSortOrder: Type.Optional(StringUnion(["descend" ,"ascend"])),
+  render: Type.Optional(Type.Any()),
```
Comment (MrWindlike) [inline src/generated/types/Table.ts]: `render` look like is an internal property for column, should not be set here.

File: src/generated/types/Table.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+export const handlerSchema = Type.Array(Type.Object({
+  componentId: Type.String(),
+  type: Type.String(),
+  method: Type.Object({
+    name: Type.String(),
+    parameters: Type.Optional(Type.Object(Type.Any()))
+  })
+}))
+
+export const ColumnSchema = Type.Object({
```
Comment (MrWindlike) [inline src/generated/types/Table.ts:14]: May support `fix` property in column?

**Review by tanbowensg — CHANGES_REQUESTED** [review]

File: src/components/Table.tsx [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    currentPage * pageSize
+  );
+
+  const clearState = () => {
+    mergeState({ selectedRows: [] });
+    mergeState({ selectedItem: undefined });
+  };
+  const inputRef = useRef(null);
+
+  for (const column of columns!) {
```
Comment (MrWindlike) [inline src/components/Table.tsx]: Change React props directly is bad. May use `columns.map` to generate a new array to do that is better.

File: src/generated/types/Table.ts (hunk 4 above)
Comment (xzdry) [reply src/generated/types/Table.ts:14]: what do you mean? modify row in table?

Comment (MrWindlike) [reply src/generated/types/Table.ts:14]: image
Sorry, that's typo. It's `fixed` property. I mean that should support columns fixed at left or right.

Comment (xzdry) [reply src/generated/types/Table.ts:14]: sure

**Pull request merged**